server/WFCORE-4291_restore_legacy_not_graceful_startup_mode.adoc

```diff
@@ -0,0 +1,68 @@
+= WFCORE-4291 - Restore legacy (not "graceful") startup mode
+:author:            Jason Lee
+:email:             jasondlee@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+
+== Overview
+
+Since Wildfly 11, the server starts up in suspended mode, thus rejecting all incoming requests, until the server has completed booting, at which point the state is set to `RUNNING` and requests are accepted. While this helps ensure that the system is in a stable state before allowing any work to happen, it does have a negative effect on some deployment scenarios reported by Wildfly users. 
+
+One example scenario is this: if Deployment A depends -- at startup -- on external configuration, fetched by an HTTP request to Deployment B, Deployment A will fail to start as the HTTP request for that configuration will be rejected by the starting server.
+
+This RFE proposes a new startup mode, `graceless`, that will allow -- for some subsystems, at least -- requests to be handled much earlier in the application server boot process. In standalone mode, the system administrator can choose to enable this startup mode for that server, and in domain mode, the server group can be configured similarly, applying the configuration choice to all servers in that domain.
+
+== Issue Metadata
+
+=== Issue
+* https://issues.jboss.org/browse/WFCORE-4291[WFCORE-4291]
+
+=== Related Issues
+* https://issues.redhat.com/browse/EAP7-1320[EAP7-1320]
+
+=== Dev Contacts
+* mailto:{email}[{author}]
+
+=== QE Contacts
+* mailto:jkasik@redhat.com[Jan Kasik]
+
+=== Testing By
+* [X] QE
+
+=== Affected Projects or Components
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+For standalone mode, this option will be exposed as a command line option only, with the addition of a `--graceless` option. Since the value is needed before standalone.xml is parsed, it must come from the command line. Given this lack of start mode persistence, system administrators managing standalone instances will need to update their system scripts (`/etc/init.d`, `sysvinit`, `systemd`, etc.)
+
+For domain mode, the configuration will be extended to allow the system administrator to configure how the Host Contoller will start the servers in the group. The `server-group` element will support a new `graceless` attribute, whose default value will be `false`. Every server in the group will be configured to start in the same manner.
+
+=== Nice-to-Have Requirements
+
+=== Non-Requirements
+
```

Comment:
How is this feature compatible with WildFly/EAP on OpenShift and current implementation of readiness/liveness probes?

Reply:
I *think* that this makes it possible for that check to return an ... erroneous error :) should the check be made during startup (i.e., the server is starting in gracless so the request is processed, but the server's not quite ready to handle it, so an error is returned). How that might affect OS I don't know. @bstansberry?

Reply:
In my opinion it should be decided if this feature should be supported on OpenShift. Right now, if it was already implemented, nothing stops user from trying to use it: They can use `JAVA_OPTS_APPEND` to add the command line option to activate "graceless" mode.

AFAIK the pod with running WildFly is added to service load balancer after it reports that it is ready (meaning all containers in it are ready). I don't know how implementing this feature will affect reporting of readiness and liveness, but if it would still work the same as it works now, it would mean, that using this for cross-server requests doesn't make sense. However it could still be useful for scenarios where deployment calls itself or other deployments on the same server inside one pod, without using OCP defined networking.

This should be IMHO cleared up in this analysis.

Reply:
I talked to Brian yesterday, and he said that the liveness/readiness checks are done over the management interface, which is not affected by this change in startup behavior, so my guess above is incorrect. We should be fine with regard to OpenShift. I can update the document to indicate that.

Reply:
Thanks @jasondlee! So if I understood it correctly the liveness/readiness check won't be affected by this feature means that inter service communication on OCP won't benefit from this. The only scenario where this will be "supported"/usable is when deployment on starting server "calls home" to itself or deployments on the same server in one pod. Am I correct with this assumption? This should be included in the analysis document in my opinion.

Reply:
That is correct.

Reply:
@jasondlee Can you please include this in non-requirements then?